### ISSUE TITLE
Allow derivations from extended public keys

### ISSUE THREAD

**Issue by yawhide** [meherett/python-hdwallet#12]
Hello again :)

It looks like HDWallet only allows deriving extended public keys when you have the private key. I believe it is possible to derive extended public keys when you do not have the private key.

I am trying to get ALL sorts of bip32/44/49/84 extended public keys that are from various cryptocurrencies (like btc, doge, ltc, dash, zcash) and then trying to derive some extended public keys from them.

I am trying to do as follows:

```python
hdwallet = HDWallet(symbol='DASH')
hdwallet.from_xpublic_key(xpublic_key='drkpRykits4vMCKnaJsr4e6z1pSAkHtWzx9ZFiZJzRaq3xo2Ug9eFVBdQ9HAMSMUehRk4ANkeyzbtEXVY91LM4RkojPzN3eTaHQLtquM6LRypUm')
addresses = []
i = 0
while i <= 2:
    hdwallet.from_path(path=f'm/0/{i}')
    addresses.append(hdwallet.address())
    hdwallet.clean_derivation()
    i += 1
print(addresses) # ['XbTRLsLqMviTsow6g4y3apHZNCc3XudymK', 'XkDJ5Vg5nJz9FSbTHxqAZ1kjmoZfNZYo86', 'XcG6n92PP9mco6hJFNfRA5KkMVSRmEETwk']
```

but I get the following exception:

```python
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/toor/coin-tracker/coin-tracker-server/env/lib/python3.6/site-packages/hdwallet/hdwallet.py", line 398, in from_path
    self._derive_key_by_index(int(index))
  File "/home/toor/coin-tracker/coin-tracker-server/env/lib/python3.6/site-packages/hdwallet/hdwallet.py", line 452, in _derive_key_by_index
    pvt_int = string_to_int(self._key.to_string())
AttributeError: 'NoneType' object has no attribute 'to_string'
``` 

thanks again!

**Comment by MarcoVHalliday** [comment]
Its good.

**Comment by meherett** [comment]
Hi @yawhide, Private key is required. You can't drive extended public keys without private key.

**Comment by yawhide** [comment]
> Hi @yawhide, Private key is required. You can't drive extended public keys without private key.

You totally can.

If you go to https://iancoleman.io/bip39/ and put in an extended public key in the text box for bip32 root key  you can generate. You can leave the seed or mnemonic phrase blank.

**Comment by yawhide** [comment]
@meherett bump :) ^

**Comment by meherett** [comment]
@yawhide It doesn't work without root xprivate key,  it's also mandatory to get addresses.

> > Hi @yawhide, Private key is required. You can't drive extended public keys without private key.
> 
> You totally can.
> 
> If you go to https://iancoleman.io/bip39/ and put in an extended public key in the text box for bip32 root key you can generate. You can leave the seed or mnemonic phrase blank.
> 
> example:
> go to iancoleman's website, put `Ltub2YKc9ECMezkgir4yAVSWZTo6xFJ3vC2SSYZj8pNJdBTFUo13HG1oZdsX9JFbt6WbMEj3dLJxkNhkHZSpZyhSKxsYeo31GUV12Kq9uW1r3YQ` as the bip32 root key and select `Litecoin` as the coin.
> 
> then click the `bip32` tab below. you can see all the addresses you can derive.

It's not working for me.

**Comment by yawhide** [comment]
try this:
`xpub68tcYNiCwm6SZktWMMse3gKmnWDjjaifGSKK92rZ2sneDnXRS8QAUN35134FW5JSXosoeXz6SU3nycd2gaGZ2iYhHMLbJR5wnAxFAiZnAir`

![iancoleman io_bip39_](https://user-images.githubusercontent.com/5014112/121527438-66300f80-c9c8-11eb-9b85-5356d8547429.png)


this has to be possible or else no service that can track your "portfolio" given an xpub would work. There is no way people are giving their xpriv to random portfolio tracking (or even blockchain explorers) websites.

Here is another website (blockchain.com). given this xpub you can see all the addresses it found with transactions https://www.blockchain.com/btc/xpub/xpub6CUGRUonZSQ4TWtTMmzXdrXDtypWKiKrhko4egpiMZbpiaQL2jkwSB1icqYh2cfDfVxdx4df189oLKnC5fSwqPfgyP3hooxujYzAu3fDVmz (xpub is in the URL)

Here is an api where you can give it an extended public key and it'll find any transactions with addresses generated from that xpub https://blockchair.com/api/docs#link_300

**Comment by meherett** [comment]
Hi @yawhide, I found it last night but the hardened derivation path is invalid for xpublic key.  So, I will release it tomorrow or today.

```python
i_str = struct.pack(">L", index)
if index & BIP32KEY_HARDEN:
   data = b"\0" + self._key.to_string() + i_str
else:
   data = unhexlify(self.public_key()) + i_str
```

https://github.com/meherett/python-hdwallet/blob/master/hdwallet/hdwallet.py#L438-L441

Thank you, bro :)

**Comment by meherett** [comment]
@yawhide now it's available on the v1.3.0 package. Just check this example https://github.com/meherett/python-hdwallet/blob/master/examples/from_root_xpublic_key.py

Thanks :)
